fearless/Common/Operation/StorageRequestFactory.swift

```diff
@@ -0,0 +1,181 @@
+    func queryItems<K, T>(engine: JSONRPCEngine,
+                          keyParams: @escaping  () throws -> [K],
+                          factory: @escaping  () throws -> RuntimeCoderFactoryProtocol,
+                          storagePath: StorageCodingPath)
+    -> CompoundOperationWrapper<[StorageResponse<T>]> where K: Encodable, T: Decodable
+
+    func queryItems<K1, K2, T>(engine: JSONRPCEngine,
+                               keyParams1: @escaping  () throws -> [K1],
+                               keyParams2: @escaping  () throws -> [K2],
+                               factory: @escaping  () throws -> RuntimeCoderFactoryProtocol,
+                               storagePath: StorageCodingPath)
+    -> CompoundOperationWrapper<[StorageResponse<T>]> where K1: Encodable, K2: Encodable, T: Decodable
+
+    func queryItems<T>(engine: JSONRPCEngine,
+                       keys: @escaping () throws -> [Data],
+                       factory: @escaping  () throws -> RuntimeCoderFactoryProtocol,
+                       storagePath: StorageCodingPath)
+    -> CompoundOperationWrapper<[StorageResponse<T>]> where T: Decodable
+}
+
+final class StorageRequestFactory: StorageRequestFactoryProtocol {
+    let remoteFactory: StorageKeyFactoryProtocol
+
+    init(remoteFactory: StorageKeyFactoryProtocol) {
+        self.remoteFactory = remoteFactory
+    }
+
+    func queryItems<T>(engine: JSONRPCEngine,
+                       keys: @escaping () throws -> [Data],
+                       factory: @escaping  () throws -> RuntimeCoderFactoryProtocol,
+                       storagePath: StorageCodingPath)
+    -> CompoundOperationWrapper<[StorageResponse<T>]> where T: Decodable {
+        let queryOperation = JSONRPCQueryOperation(engine: engine,
+                                                   method: RPCMethod.queryStorageAt)
+        queryOperation.configurationBlock = {
+            do {
+                let keys = try keys().map { $0.toHex(includePrefix: true) }
+                queryOperation.parameters = [keys]
+            } catch {
+                queryOperation.result = .failure(error)
+            }
+        }
+
+        let decodingOperation = StorageDecodingListOperation<T>(path: storagePath)
+        decodingOperation.configurationBlock = {
+            do {
+                let result = try queryOperation.extractNoCancellableResultData()
+
+                decodingOperation.codingFactory = try factory()
+
+                decodingOperation.dataList = result
+                    .flatMap({ StorageUpdateData(update: $0).changes })
+                    .compactMap { $0.value }
+            } catch {
+                decodingOperation.result = .failure(error)
+            }
+        }
+
+        decodingOperation.addDependency(queryOperation)
+
+        let mapOperation = ClosureOperation<[StorageResponse<T>]> {
+            let result = try queryOperation.extractNoCancellableResultData()
+            let keysData = result
```

Comment:
result.flatMap({ StorageUpdateData(update: $0).changes }) is computed 3 times. Suggest define 
```suggestion
            let resultChangesData = result.flatMap { StorageUpdateData(update: $0).changes }
```

Reply:
Fixed